### PR TITLE
txn: change memory pessimsitic lock to btree map and support scan

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What is changed and how it works?

Issue Number: Ref #15066 #16158 

What's Changed:

```commit-message
Change in-memory pessimsitic locks from hash map to btree map, support collecting pessimistic locks for scan lock command.
Then:
1. GC could collect expired pessimistic locks.
2. Pessimistic rollback could use read scan first and then clean up expired pessimistic locks at one time.
```

# Benchmark Test

## Local Sysbench
From the sysbench `oltp_write_only` tests, there are not much performance difference
PR:
Throughput:
    events/s (eps):                      5082.2634
    time elapsed:                        600.0055s
    total number of events:              3049386

Master:
Throughput:
    events/s (eps):                      4971.9635
    time elapsed:                        600.0060s
    total number of events:              2983208

## Heavy Lock Workload
Besides @zyguan has helped to verify that in lock heavy workloads there could be a **3% regression**, see the comments below for details. 

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory

### Release note

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/txn_ext.rs
Comment (ekexium) [inline components/raftstore/src/store/txn_ext.rs:247]: `drain_filter` has been renamed to `extract_if` in newer versions. I see BTreeMap has `drain_filter` in the version TiKV uses.

File: components/raftstore/src/store/txn_ext.rs
Comment (ekexium) [inline components/raftstore/src/store/txn_ext.rs:250]: Just curious, why not `key.as_encoded().as_slice()`

File: components/raftstore/src/store/txn_ext.rs
Comment (cfzjywxk) [reply components/raftstore/src/store/txn_ext.rs:247]: `extract_if` is still a nightly-only experimental api(https://github.com/rust-lang/rust/issues/70530) and seems unstable. like `drain_filter` maybe it's difficult to implement such thing on the `BtreeMap`.

File: components/raftstore/src/store/txn_ext.rs
Comment (cfzjywxk) [reply components/raftstore/src/store/txn_ext.rs:250]: `key.as_encoded().as_slice()` is a better way.

File: src/storage/mod.rs
Comment (MyonKeminta) [inline src/storage/mod.rs:1486]: Sadly we cannot merge the two results by simply concatenating them. TiDB's GCWorker will assume that when a scan_lock invocation returns a (non-empty) list of locks, then there won't be any other unknown stale locks in range `[start_key, locks[len(locks)-1].key]`. So if the count of locks exceeds the `limit` param of the request, TiDB can then perform the next scan with start_key =  `locks[len(locks-1)].key`.
In other words, scan_locks should guarantee that it returns the minimal N locks in the range where N = min(number  of locks in range, limit).

Another problem is that TiDB determines whether the scanning on the range is finished by checking whether the returned locks is less than limit. So I think it's a bad idea to return more than `limit` locks. Here the code seems 
 to possibly return at most 2*limit locks (memlock*limit + persistedlock*limit). Also, it's confusing not to follow how it's actually requested by specifying `limit`.

I think you can simply sort the vector `locks` and keep only the first `limit` items if it's too many.
The GCWorker doesn't rely on the order of locks in the response (if I remember it correctly), but I still suggest to keep the order to keep the behavior consistent.

File: components/raftstore/src/store/txn_ext.rs
Comment (MyonKeminta) [reply components/raftstore/src/store/txn_ext.rs:247]: Considering that:
* TiKV has usages of `Vec` and `HashMap`'s `extract_if` method though they are also exparimental
* Currently `retain` is actually implemented by `extract_if`
Actually I think perhaps introducing an usage of another experimental API might be an acceptable choice 🤔

File: components/raftstore/src/store/txn_ext.rs
Comment (MyonKeminta) [inline components/raftstore/src/store/txn_ext.rs:250]: `key` is internally in encoded form, and `as_encoded` gets an reference to it. It's not very necessary to emphasize it's encoded here, as all other keys being used here are encoded too.

File: components/raftstore/src/store/txn_ext.rs
Comment (MyonKeminta) [inline components/raftstore/src/store/txn_ext.rs:301]: You can check `locks.len() >= limit`

**Review by zyguan — COMMENTED** [review]
I've run the following lock heavy workload with command options: `--batch-updates=1 --batch-size=1000 --mode=select-for-update --sequential=on --lock-only=on`. 

<details>
<summary>code</summary>

```lua
require("oltp_common")

sysbench.cmdline.options.batch_updates = {
    'Number of batch updates per transaction', 1}
sysbench.cmdline.options.batch_size = {
    'Batch size (number of records per update)', 10}
sysbench.cmdline.options.mode = {
    'Update mode {default, update-index, select-for-update}', 'default'}
sysbench.cmdline.options.lock_only = {
    'Only lock rows', false}
sysbench.cmdline.options.sequential = {
    'Update sequentially in one batch', false}

function prepare_commit()
    if sysbench.opt.lock_only then
        stmt.commit = con:prepare("ROLLBACK")
    else
        stmt.commit = con:prepare("COMMIT")
    end
end

function prepare_statements()
    if not sysbench.opt.skip_trx then
        prepare_begin()
        prepare_commit()
    end

    local xs = {}
    local stmt_pattern = ''
    for i = 1, sysbench.opt.batch_size do table.insert(xs, "?") end
    if sysbench.opt.mode == 'select-for-update' then
        stmt_pattern = "SELECT c FROM sbtest%u WHERE id in ("..table.concat(xs, ",")..") FOR UPDATE"
    elseif sysbench.mode == 'update-index' then
        stmt_pattern = "UPDATE sbtest%u SET k=k+1 WHERE id in ("..table.concat(xs, ",")..")"
    else
        stmt_pattern = "UPDATE sbtest%u SET c=REVERSE(c) WHERE id in ("..table.concat(xs, ",")..")"
    end
    for t = 1, sysbench.opt.tables do
        stmt[t].batch_update = con:prepare(string.format(stmt_pattern, t))
        param[t].batch_update = {}
        for p = 1, sysbench.opt.batch_size do
            param[t].batch_update[p] = stmt[t].batch_update:bind_create(sysbench.sql.type.INT)
        end
        stmt[t].batch_update:bind_param(unpack(param[t].batch_update))
     end
end

function event()
    if not sysbench.opt.skip_trx then
        begin()
    end

    local tnum = sysbench.rand.uniform(1, sysbench.opt.tables)
    for i = 1, sysbench.opt.batch_updates do
        if sysbench.opt.sequential then
            local base = sysbench.rand.default(1, math.max(1, sysbench.opt.table_size-sysbench.opt.batch_size))
            for j = 1, sysbench.opt.batch_size do
                param[tnum].batch_update[j]:set(base + j - 1)
            end
        else
            for j = 1, sysbench.opt.batch_size do
                param[tnum].batch_update[j]:set(sysbench.rand.default(1, sysbench.opt.table_size))
            end
        end
        stmt[tnum].batch_update:execute()
    end

    if not sysbench.opt.skip_trx then
        commit()
    end

    check_reconnect()
end
```

</details>

The results show that there is less than a 3% regression compare to the baseline. I think it's acceptable.

File: components/raftstore/src/store/txn_ext.rs
Comment (cfzjywxk) [reply components/raftstore/src/store/txn_ext.rs:247]: @MyonKeminta 
Sorry I did not find the `#![feature(btree_extract_if)]` related usages in the repo, and the `retaint` is internal implemented by `DrainFilterInner`.
For `HashMap` it is much easier to implement drain filter.

Comment (MyonKeminta) [reply components/raftstore/src/store/txn_ext.rs:247]: I didn't mean that there are any existing use of `btree_extract_if`, but just considering whether it's acceptable to introduce it.

> and the retaint is internal implemented by DrainFilterInner.

Ok.. sorry, guess I was reading a different version

File: src/storage/mod.rs
Comment (cfzjywxk) [reply src/storage/mod.rs]: Thanks for the reminding! The lock result merging logic is added, the `merge_by` function is used to merge the two sorted vectors, PTAL.

File: Cargo.toml
Comment (MyonKeminta) [inline Cargo.toml:1]: Better revert this file before merging

File: src/storage/mvcc/reader/reader.rs
Comment (MyonKeminta) [inline src/storage/mvcc/reader/reader.rs:290]: Should we consider adding a metric here to show how long the rwlock is hold? I think it might be helpful to see how much it affect the performance of normal queries.

**Review by MyonKeminta — COMMENTED** [review]
Rest LGTM

File: tests/integrations/server/kv_service.rs
Comment (cfzjywxk) [reply tests/integrations/server/kv_service.rs:2777]: Increasing the `in-memory` lock size limit may help I think.

File: tests/integrations/server/kv_service.rs
Comment (cfzjywxk) [reply tests/integrations/server/kv_service.rs:2794]: Good catch, I'll add some tests when the pessimsitic locks are written both in memory and in the lock cf.

File: tests/integrations/server/kv_service.rs
Comment (MyonKeminta) [reply tests/integrations/server/kv_service.rs:2777]: I mean, can we try to check whether this test code writes pessimistic locks to memory or storage? In case there are any missing configuration which leads to the locks still being written to storage, then the test would become ineffective

File: src/storage/mod.rs
Comment (MyonKeminta) [inline src/storage/mod.rs:7422]: I just noticed that the test storage seems not enabling in-memory pessimistic locks by default.
How about testing both enabling and disabling?

**Review by zyguan — APPROVED** [review]

File: tests/integrations/server/kv_service.rs
Comment (cfzjywxk) [reply tests/integrations/server/kv_service.rs:2782]: There seems no direct way to check the memory lock status of the started cluster, some engine checks are added at L2804 to verify if the locks exist in the lock cf, PTAL.

File: src/storage/mod.rs
Comment (cfzjywxk) [reply src/storage/mod.rs]: Thanks for the reminding, both are added and so is the memory lock checking, PTAL.

**Review by MyonKeminta — APPROVED** [review]

**Pull request merged**
